### ISSUE TITLE
The process instance is stuck if the output collection is modified

### ISSUE THREAD

**Issue by saig0** [camunda/zeebe#6953]
**Describe the bug**
If the output collection variable is modified then the process instance is stuck in the multi-instance subprocess.

The sequential/parallel multi-instance subprocess iterates over an input collection and collects the results in an output collection. If the value of the output collection variable is modified (e.g. change the type of the value or reduce the size of the collection) then the inner instance can't be completed. 

The process instance doesn't continue (i.e. it is added to the blacklist). No incidents are created. 

![image](https://user-images.githubusercontent.com/4305769/116963297-c84e7400-aca8-11eb-87aa-879d980b4868.png)

**To Reproduce**
1. deploy a process with a sequential/parallel multi-instance subprocess that defines
  * an input collection `= items`
  * an input element `item`
  * an output collection `results`
  * an output element `= r`
2. create an instance with variables `{"items":[1,2,3]}`
3. await until the service task in the subprocess is activated
4. set the value of the variable `results` to `123` (some value that is not a list)
5. complete the job
6. verify that the process instance doesn't continue 

**Expected behavior**
It creates an incident when completing the inner instance.
I can resolve the incident and the process instance continues.

**Log/Stacktrace**

```
MsgpackReaderException: Unable to determine array type, found unknown header byte 0x7b at reader offset 0
```

<details><summary>Full Stacktrace</summary>
 <p>

```
io.zeebe.msgpack.spec.MsgpackReaderException: Unable to determine array type, found unknown header byte 0x7b at reader offset 0
at io.zeebe.msgpack.spec.MsgPackReader.exceptionOnUnknownHeader (MsgPackReader.java:474)
at io.zeebe.msgpack.spec.MsgPackReader.readArrayHeader (MsgPackReader.java:107)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.insertAt (MultiInstanceBodyProcessor.java:372)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.lambda$updateOutputCollection$14 (MultiInstanceBodyProcessor.java:355)
at io.zeebe.util.Either$Right.map (Either.java:213)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.updateOutputCollection (MultiInstanceBodyProcessor.java:346)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.lambda$updateOutputCollection$13 (MultiInstanceBodyProcessor.java:332)
at java.util.Optional.map (Unknown Source)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.updateOutputCollection (MultiInstanceBodyProcessor.java:330)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.beforeExecutionPathCompleted (MultiInstanceBodyProcessor.java:150)
at io.zeebe.engine.processing.bpmn.container.MultiInstanceBodyProcessor.beforeExecutionPathCompleted (MultiInstanceBodyProcessor.java:35)
at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.lambda$beforeExecutionPathCompleted$2 (BpmnStateTransitionBehavior.java:406)
at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.invokeElementContainerIfPresent (BpmnStateTransitionBehavior.java:489)
at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.beforeExecutionPathCompleted (BpmnStateTransitionBehavior.java:402)
at io.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior.transitionToCompletedWithParentNotification (BpmnStateTransitionBehavior.java:167)
at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.lambda$onComplete$3 (SubProcessProcessor.java:78)
at io.zeebe.util.Either$Right.ifRightOrLeft (Either.java:238)
at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.onComplete (SubProcessProcessor.java:74)
at io.zeebe.engine.processing.bpmn.container.SubProcessProcessor.onComplete (SubProcessProcessor.java:23)
at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processEvent (BpmnStreamProcessor.java:213)
at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.lambda$processRecord$1 (BpmnStreamProcessor.java:134)
at io.zeebe.util.Either$Right.ifRightOrLeft (Either.java:238)
at io.zeebe.engine.processing.bpmn.BpmnStreamProcessor.processRecord (BpmnStreamProcessor.java:131)
at io.zeebe.engine.processing.streamprocessor.TypedRecordProcessor.processRecord (TypedRecordProcessor.java:54)
at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.lambda$processInTransaction$3 (ProcessingStateMachine.java:296)
at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run (ZeebeTransaction.java:84)
at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.processInTransaction (ProcessingStateMachine.java:286)
at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.processEvent (ProcessingStateMachine.java:254)
at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.tryToReadNextEvent (ProcessingStateMachine.java:219)
at io.zeebe.engine.processing.streamprocessor.ProcessingStateMachine.readNextEvent (ProcessingStateMachine.java:210)
at io.zeebe.util.sched.ActorJob.invoke (ActorJob.java:73)
at io.zeebe.util.sched.ActorJob.execute (ActorJob.java:39)
at io.zeebe.util.sched.ActorTask.execute (ActorTask.java:122)
at io.zeebe.util.sched.ActorThread.executeCurrentTask (ActorThread.java:94)
at io.zeebe.util.sched.ActorThread.doWork (ActorThread.java:78)
at io.zeebe.util.sched.ActorThread.run (ActorThread.java:191)
```

</p>
</details>

**Environment:**
 - OS: Camunda Cloud
 - Zeebe Version: `1.0.0-alpha7`
 - Configuration:

**Comment by oleschoenburg** [comment]
I'd like to raise awareness for this issue again. It's still occurring on `8.0.1`: https://console.cloud.google.com/errors/detail/CPH5wMiVr_Xy5wE;service=zeebe;time=P7D?project=camunda-cloud-240911

**Comment by saig0** [comment]
This issue is fixed by #9178 in version `8.0.2` :tada:
